### PR TITLE
Add validation tests for Multi-Draw Indirect

### PR DESCRIPTION
**Pull request by beaufortfrancois**
This PR adds validation tests for the experimental Multi-Draw Indirect feature that can be tested in Chrome Canary behind the "Unsafe WebGPU support" flag at `chrome://flag#enable-unsafe-webgpu`

Issue: https://github.com/gpuweb/cts/issues/3961

image

<hr>

**Requirements for PR author:**

- [ ] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [ ] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [ ] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .params(u =>
+    u.expandWithParams(() => {
+      return [
+        // In bounds
+        { bufferSize: 4, indirectOffset: 0, _valid: true },
```
Comment (Sirtsu55) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:230]: nit: Could we combine these test cases with the ones below as a variable? If I didn't miss anything these cases seem similar with small differences, and combining them would be cleaner/more readable.

**Review by Sirtsu55 — COMMENTED** [review]
Apart from the stylistic nitpick, LGTM.

**Comment by Sirtsu55** [comment]
FYI @beaufortfrancois , I'm working on #3962 to add more rendering tests. Should we incrementally add rendering tests to this branch or should submit all of the rendering tests at once?

**Comment by beaufortfrancois** [comment]
> FYI @beaufortfrancois , I'm working on #3962 to add more rendering tests. Should we incrementally add rendering tests to this branch or should submit all of the rendering tests at once?

I'll let @kainino0x weigh in as I don't have strong preferences.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+  .desc(
+    `
+Tests multi indirect draw calls with various offsets and buffer sizes.
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .params(u =>
+    u.expandWithParams(() => {
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:227]: nit: use `.combineWithParams([...])`

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+  .desc(
+    `
+Tests multi indexed indirect draw calls with various offsets and buffer sizes.
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .params(u =>
+    u.expandWithParams(() => {
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:334]: nit: use `.combineWithParams([...])`

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 1 above)
Comment (kainino0x) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:230]: +1, if possible store this array as a variable and use it twice. A pattern we use elsewhere to make this work is instead of having `bufferSize` as a parameter, have something like `bufferExtraSize` which has negative, zero, and positive cases (`4 + (-1) = 3` for indirect, `5 + (-1) = 4` for indexed indirect)

And if there are any cases that are specific to one or the other then you can do:

```ts
const kIndirectOffsetOOBCommonCases = [
  /* ... */
] as const;

/* ... */
  .combineWithParams([
    ...kIndirectOffsetOOBCommonCases,
    /* any additional cases which are not common between the two */
  ])
```

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 3 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 2 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 1 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: I've hopefully addressed your feedback in latest patch. Please have a look!

Comment (Sirtsu55) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Sorry for the late response, but the patch looks good to me!

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+interface IndirectOffsetOobCase {
+  indirectOffset: number;
+  bufferSize: number;
+  useDrawCountBuffer?: boolean;
+  drawCountOffset?: number;
+  maxDrawCount?: number;
+  _valid: boolean;
+}
+
+g.test('indirect_offset_oob')
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:234]: Test name implies it only tests OOB on the indirectBuffer, but is also doing a bunch of stuff with drawCountBuffer that is hard to separate in my head. Can there be two tests:
- `indirectBuffer_OOB` (just a single boolean param for whether it uses a drawCountBuffer or not)
- `drawCountBuffer_OOB`
?

File: src/webgpu/api/validation/encoding/encoder_open_state.spec.ts [hunk 5]
```diff
@@ -101,6 +105,8 @@ const kRenderPassEncoderCommandInfo: {
   pushDebugGroup: {},
   popDebugGroup: {},
   insertDebugMarker: {},
+  multiDrawIndirect: {},
+  multiDrawIndexedIndirect: {},
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/encoder_open_state.spec.ts]: Done

File: src/webgpu/api/validation/encoding/encoder_open_state.spec.ts [hunk 6]
```diff
@@ -414,6 +426,18 @@ g.test('render_pass_commands')
...
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (renderPass as any).multiDrawIndirect(buffer, 0, 1);
+          }
+          break;
+        case 'multiDrawIndexedIndirect':
+          {
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (renderPass as any).multiDrawIndexedIndirect(buffer, 0, 1);
+          }
+          break;
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/encoder_open_state.spec.ts]: Done

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+
+  .paramsSubcasesOnly(
+    kIndirectMultiDrawTestParams
+      .combine('indirectState', kResourceStates)
+      .combine('drawCountState', kResourceStates)
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:43]: LGTM. Thanks!

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+    const submitShouldSucceedIfValid =
+      indirectState !== 'destroyed' && (!drawCountBuffer || drawCountState !== 'destroyed');
+    validateFinishAndSubmit(shouldBeValid, submitShouldSucceedIfValid);
+  });
+
+g.test('buffers,device_mismatch')
+  .desc(
+    'Tests multiDraw(Indexed)Indirect cannot be called with buffers created from another device'
+  )
+  .paramsSubcasesOnly(kIndirectMultiDrawTestParams.combine('mismatched', [true, false]))
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+Tests indirect and draw count offsets must be a multiple of 4.
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
+  .paramsSubcasesOnly(
+    kIndirectMultiDrawTestParams
+      .combine('indirectOffset', [0, 2, 4] as const)
+      .combine('drawCountOffset', [0, 2, 4] as const)
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done with `combineWithParams`

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+  - (kMaxUnsignedLongLongValue, min size)
+  - (0, min size) with drawCountOffset = kMaxUnsignedLongLongValue
+  - (0, min size) with maxDrawCount = kMaxUnsignedLongValue
+- min size = indirect draw parameters size
+- x =(multiDrawIndirect, multiDrawIndexedIndirect)
+  `
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('chromium-experimental-multi-draw-indirect' as GPUFeatureName);
+  })
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+  - (0, min size - min alignment)
+  - (min alignment +/- 1, min size + min alignment)
+  - (0, min size + min alignment) with drawCountBuffer and drawCountOffset not a multiple of 4
+  - (1, min size) index too big
+  - (min size + min alignment, min size) index past buffer
+  - (0, min size) with maxDrawCount = 2
+  - (0, min size) with drawCountOffset = min size
+  - (kMaxUnsignedLongLongValue, min size)
+  - (0, min size) with drawCountOffset = kMaxUnsignedLongLongValue
+  - (0, min size) with maxDrawCount = kMaxUnsignedLongValue
```
[the remark this answers is not in the document]
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: I think the comment above each test such as `// Out of bounds, buffer too small` serve this purpose. I've removed this list then.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 4 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: I've renamed this test as the implementation of two tests would be identical.

**Review by beaufortfrancois — COMMENTED** [review]
Thanks for the thorough review @kainino0x!

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 12]
```diff
@@ -112,7 +128,7 @@ g.test('buffers,device_mismatch')
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       (encoder as any).multiDrawIndirect(indirectBuffer, 0, 1, drawCountBuffer);
     }
-    validateFinish(!mismatched);
+    validateFinish(!indirectMismatched && (!useDrawCountBuffer || !drawCountMismatched));
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:131]: Ah I forgot also that drawCountMismatched wouldn't matter if !useDrawCountBuffer. That useless case should ideally be filtered out as well.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 4 above)
Comment (kainino0x) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: I'd prefer if the two were totally separate, as the validation for OOB on each of the two buffers is totally orthogonal. For offsets_alignment it's fine because the list of cases is small, but this list of cases is large and needs to test all the corner cases of two separate checks. Reviewing the list to make sure it covers all of those cases is difficult.

If I'm misunderstanding and they're not orthogonal then this is fine, I just still will need to review the list of cases more closely.

Comment (kainino0x) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: (Also the other tests use kIndirectMultiDrawTestParams so if they were split I think they would have a lot of duplicated control cases; that's not the case here)

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 12 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts:131]: Done.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts (hunk 4 above)
Comment (beaufortfrancois) [reply src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: Done.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+        drawCountBuffer,
+        drawCountOffset
+      );
+    }
+
+    // We need to figure out if https://github.com/gpuweb/gpuweb/pull/2315/files#r1773031950 applies.
+    validateFinish(indirectOffset % 4 === 0 && (!useDrawCountBuffer || drawCountOffset % 4 === 0));
+  });
+
+g.test('indirect_buffer_OOB')
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: nit: Realized this also validates alignment in addition to OOB, so more generally:
```suggestion
g.test('indirectBuffer_range')
```

also `indirectBuffer` because it refers to the argument called `indirectBuffer`.

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+      (encoder as any).multiDrawIndexedIndirect(indirectBuffer, indirectOffset, maxDrawCount);
+    } else {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (encoder as any).multiDrawIndirect(indirectBuffer, indirectOffset, maxDrawCount);
+    }
+
+    validateFinish(_valid);
+  });
+
+g.test('draw_count_buffer_OOB')
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]:
```suggestion
g.test('drawCountBuffer_range')
```

File: src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+g.test('indirect_buffer_OOB')
+  .desc(
+    `
+Tests multi indirect draw calls with various indirect offsets and buffer sizes without draw count buffer.
+`
+  )
+  .paramsSubcasesOnly(u =>
+    u.combine('indexed', [true, false] as const).expandWithParams(p => {
+      const indirectParamsSize = p.indexed ? 20 : 16;
+      return [
```
Comment (kainino0x) [inline src/webgpu/api/validation/encoding/cmds/render/indirect_multi_draw.spec.ts]: It is hard to tell reading this list which boundary conditions are checked correctly. Generally we want pairs of control case and error case where the error case is minimally-wrong: just slightly too small, or just slightly too unaligned.

I've pushed commits that refactor this a bit (and address my other comments), PTAL.

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
PTAL. If my changes look good (they pass in Chrome), you can go ahead and land the PR.

**Comment by beaufortfrancois** [comment]
I've checked locally and all tests pass in Chrome! Thank you @kainino0x

**Comment by kainino0x** [comment]
Sorry, I wrote that unclearly :)
what I meant was, "They already pass in chrome, so if they look good to you, you can land it"
